gobblin-runtime/src/main/java/gobblin/runtime/commit/FsCommitSequenceStore.java

```diff
@@ -0,0 +1,142 @@
+import com.google.common.base.Strings;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.Lists;
+import com.google.gson.Gson;
+
+import gobblin.commit.CommitSequence;
+import gobblin.commit.CommitSequenceStore;
+import gobblin.commit.CommitStep;
+import gobblin.configuration.ConfigurationKeys;
+import gobblin.util.HadoopUtils;
+import gobblin.util.filters.HiddenFilter;
+import gobblin.util.io.GsonInterfaceAdapter;
+
+
+/**
+ * An implementation of {@link CommitSequenceStore} backed by a {@link FileSystem}.
+ *
+ * <p>
+ *    This implementation serializes a {@link CommitSequence} along with all its {@link CommitStep}s into a
+ *    JSON string using {@link Gson}. Thus it requires that all {@link CommitStep}s can be serialized
+ *    and deserialized with {@link Gson}.
+ * </p>
+ *
+ * @author ziliu
+ */
+public class FsCommitSequenceStore implements CommitSequenceStore {
+
+  public static final String COMMIT_SEQUENCE_STORE_FS_URI = "commit.sequence.store.fs.uri";
+  public static final String COMMIT_SEQUENCE_STORE_DIR = "commit.sequence.store.dir";
+
+  private static final String DEFAULT_DATASET_URN = "default_dataset_urn";
+  private static final Gson GSON = GsonInterfaceAdapter.getGson(CommitStep.class);
+
+  private final FileSystem fs;
+  private final Path rootPath;
+
+  public FsCommitSequenceStore(FileSystem fs, Path rootPath) {
+    this.fs = fs;
+    this.rootPath = rootPath;
+  }
+
+  @Override
+  public boolean exists(String jobName) throws IOException {
+    Path jobPath = new Path(this.rootPath, jobName);
+    return this.fs.exists(jobPath);
+  }
+
+  @Override
+  public boolean exists(String jobName, String datasetUrn) throws IOException {
+    Path datasetPath = new Path(new Path(this.rootPath, jobName), sanitizeDatasetUrn(datasetUrn));
+    return this.fs.exists(datasetPath);
+  }
+
+  @Override
+  public void delete(String jobName) throws IOException {
+    Path jobPath = new Path(this.rootPath, jobName);
+    HadoopUtils.deletePathAndEmptyAncestors(this.fs, jobPath, true);
+  }
+
+  @Override
+  public void delete(String jobName, String datasetUrn) throws IOException {
+    Path jobPath = new Path(this.rootPath, jobName);
+    Path datasetPath = new Path(jobPath, sanitizeDatasetUrn(datasetUrn));
+    HadoopUtils.deletePathAndEmptyAncestors(this.fs, datasetPath, true);
+  }
+
+  @Override
+  public void put(String jobName, String datasetUrn, CommitSequence commitSequence) throws IOException {
+    datasetUrn = sanitizeDatasetUrn(datasetUrn);
+    if (exists(jobName, datasetUrn)) {
+      throw new IOException(String.format("CommitSequence already exists for job %s, dataset %s", jobName, datasetUrn));
+    }
+
+    Path jobPath = new Path(this.rootPath, jobName);
+    this.fs.mkdirs(jobPath);
+
+    Path datasetPath = new Path(jobPath, datasetUrn);
+    try (DataOutputStream dos = this.fs.create(datasetPath)) {
+      dos.writeBytes(GSON.toJson(commitSequence));
```

Comment:
Yes, both existing `CommitStep` implementations are serializable. Non-serializable fields are marked `transient` and are instantiated from the state during deserialization.